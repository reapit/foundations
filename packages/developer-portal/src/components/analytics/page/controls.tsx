import React, { FC, SetStateAction, Dispatch, useMemo } from 'react'
import {
  BodyText,
  elBorderRadius,
  elFadeIn,
  elMb3,
  elMb7,
  elWFull,
  Icon,
  InputGroup,
  Label,
  Select,
  SmallText,
} from '@reapit/elements'
import { ControlsContainer, inputFullWidth, overflowHidden, visiblyHidden } from './__styles__'
import { cx } from '@linaria/core'
import { Marketplace } from '@reapit/foundations-ts-definitions'
import { useReapitGet } from '@reapit/use-reapit-data'
import { GetActionNames, getActions } from '@reapit/use-reapit-data'
import { reapitConnectBrowserSession } from '../../../core/connect-session'
import { useReapitConnect } from '@reapit/connect-session'
import { AnalyticsFilterState, useAnalyticsState } from '../state/use-analytics-state'
import { useForm } from 'react-hook-form'
import { useLocation } from 'react-router'
import Routes from '../../../constants/routes'
import dayjs from 'dayjs'
import { MultiSelectOption } from '@reapit/elements'

export const handleFormChange =
  (setAnalyticsFilterState: Dispatch<SetStateAction<AnalyticsFilterState>>) => (values: AnalyticsFilterState) => {
    setAnalyticsFilterState(values)
  }

export const handleInstallationsToOptions = (installations?: Marketplace.InstallationModel[]) => () => {
  if (!installations) return []

  const customers = installations.map(({ customerName, client, officeGroupName }) => {
    const offGroupName = officeGroupName ? ` - ${officeGroupName} ` : ''
    return JSON.stringify({
      name: `${customerName} ${offGroupName} (${client})`,
      value: client ?? '',
    })
  })

  const options: MultiSelectOption[] = [...new Set(customers)].map((customer) => JSON.parse(customer))

  return options
}

export const Controls: FC = () => {
  const location = useLocation()
  const { connectSession } = useReapitConnect(reapitConnectBrowserSession)
  const { setAnalyticsFilterState, analyticsFilterState, analyticsDataState } = useAnalyticsState()
  const { pathname } = location
  const { appId, dateTo, monthTo, dateFrom, monthFrom } = analyticsFilterState
  const { apps } = analyticsDataState
  const developerId = connectSession?.loginIdentity.developerId
  const appQuery = appId ? { appId } : {}
  const isCallsInstallsPage = pathname === Routes.ANALYTICS_API_CALLS || pathname === Routes.ANALYTICS_INSTALLATIONS
  const isCostsPage = pathname === Routes.ANALYTICS_COSTS
  const isCalcPage = pathname === Routes.ANALYTICS_COST_CALCULATOR

  const [installations] = useReapitGet<Marketplace.InstallationModelPagedResult>({
    reapitConnectBrowserSession,
    action: getActions[GetActionNames.getInstallations],
    queryParams: { developerId, pageSize: 999, isInstalled: true, ...appQuery },
    fetchWhenTrue: [developerId],
  })

  const { register, handleSubmit } = useForm<AnalyticsFilterState>({
    mode: 'onChange',
    defaultValues: analyticsFilterState,
  })

  const options = useMemo(handleInstallationsToOptions(installations?.data), [installations])

  return (
    <div className={cx(isCalcPage && visiblyHidden, elFadeIn, elMb7)}>
      <Icon className={cx(isCalcPage && visiblyHidden, elMb3)} icon="crmInfographic" iconSize="large" />
      <BodyText className={cx(isCalcPage && visiblyHidden)}>Filter By</BodyText>
      <SmallText hasGreyText>
        Please use the filters below to search. Please note, we only hold transactional records for the last 12 months.
      </SmallText>
      <div className={cx(elBorderRadius, overflowHidden)}>
        <form onChange={handleSubmit(handleFormChange(setAnalyticsFilterState))}>
          <ControlsContainer className={cx((isCostsPage || isCalcPage) && visiblyHidden)}>
            <InputGroup
              className={inputFullWidth}
              {...register('dateFrom')}
              type="date"
              label="Date From"
              min={dayjs(dateTo).subtract(12, 'months').format('YYYY-MM-DD')}
              max={dayjs(dateTo).format('YYYY-MM-DD')}
            />
          </ControlsContainer>
          <ControlsContainer className={cx((isCostsPage || isCalcPage) && visiblyHidden)}>
            <InputGroup
              className={inputFullWidth}
              {...register('dateTo')}
              type="date"
              label="Date To"
              min={dayjs(dateFrom).format('YYYY-MM-DD')}
              max={dayjs().format('YYYY-MM-DD')}
            />
          </ControlsContainer>
          <ControlsContainer className={cx((isCallsInstallsPage || isCalcPage) && visiblyHidden)}>
            <InputGroup
              className={inputFullWidth}
              {...register('monthFrom')}
              type="month"
              label="Month From"
              min={dayjs(monthTo).subtract(12, 'months').format('YYYY-MM')}
              max={dayjs(monthTo).format('YYYY-MM')}
            />
          </ControlsContainer>
          <ControlsContainer className={cx((isCallsInstallsPage || isCalcPage) && visiblyHidden)}>
            <InputGroup
              className={inputFullWidth}
              {...register('monthTo')}
              type="month"
              label="Month To"
              min={dayjs(monthFrom).format('YYYY-MM')}
              max={dayjs().format('YYYY-MM')}
            />
          </ControlsContainer>
          <ControlsContainer className={cx(isCalcPage && visiblyHidden)}>
            <InputGroup>
              <Select className={elWFull} {...register('appId')}>
                <option key="default-option" value="">
                  None selected
                </option>
                {apps?.data?.map(({ name, id }: Marketplace.AppSummaryModel) => (
                  <option key={id} value={id}>
                    {name}
                  </option>
                ))}
              </Select>
              <Label>App</Label>
            </InputGroup>
          </ControlsContainer>
          <ControlsContainer className={cx(isCalcPage && visiblyHidden)}>
            <InputGroup>
              <Select className={elWFull} {...register('clientId')}>
                <option key="default-option" value="">
                  None selected
                </option>
                {options.map(({ name, value }) => (
                  <option key={name} value={value}>
                    {name}
                  </option>
                ))}
              </Select>
              <Label>Client</Label>
            </InputGroup>
          </ControlsContainer>
        </form>
      </div>
    </div>
  )
}
