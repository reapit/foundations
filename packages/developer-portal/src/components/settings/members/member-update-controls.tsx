import React, { FC, useEffect } from 'react'
import { Button, ButtonGroup } from '@reapit/elements'
import { SendFunction, useReapitUpdate } from '@reapit/use-reapit-data'
import { InviteMemberModel, MemberModel, UpdateMemberModel } from '@reapit/foundations-ts-definitions'
import { reapitConnectBrowserSession } from '../../../core/connect-session'
import { UpdateActionNames, updateActions } from '@reapit/use-reapit-data'
import { ReapitConnectSession } from '@reapit/connect-session'
import { useReapitConnect } from '@reapit/connect-session'

export interface MemberUpdateControlsProps {
  member: MemberModel
  refreshMembers: () => void
}

export const handleRefreshMembers =
  (
    refreshMembers: () => void,
    updateMemberSuccess?: boolean,
    deleteMemberSuccess?: boolean,
    reinviteMemberSuccess?: boolean,
  ) =>
  () => {
    if (updateMemberSuccess || deleteMemberSuccess || reinviteMemberSuccess) {
      refreshMembers()
    }
  }

export const handleUpdateMember =
  (updateMember: SendFunction<UpdateMemberModel, boolean>, memberUpdate: UpdateMemberModel) => () => {
    updateMember(memberUpdate)
  }

export const handleDeleteMember = (deleteMember: SendFunction<undefined, boolean>) => () => {
  deleteMember(undefined)
}

export const handleReinviteMember =
  (
    reinviteMember: SendFunction<InviteMemberModel, boolean>,
    member: MemberModel,
    connectSession: ReapitConnectSession | null,
  ) =>
  () => {
    const { name, jobTitle, email } = member
    const sender = connectSession?.loginIdentity.email
    const orgName = connectSession?.loginIdentity.orgName
    reinviteMember({
      name,
      jobTitle,
      email,
      sender,
      message: `Resending your invite to the ${orgName} in the Reapit Developer Portal`,
    })
  }

export const MemberUpdateControls: FC<MemberUpdateControlsProps> = ({ member, refreshMembers }) => {
  const { connectSession } = useReapitConnect(reapitConnectBrowserSession)
  const currentMemberEmail = connectSession?.loginIdentity.email

  const [, memberUpdating, updateMember, updateMemberSuccess] = useReapitUpdate<UpdateMemberModel, boolean>({
    reapitConnectBrowserSession,
    action: updateActions[UpdateActionNames.updateMember],
    method: 'PUT',
    uriParams: {
      developerId: member.developerId,
      memberId: member.id,
    },
  })

  const [, memberDeleting, deleteMember, deleteMemberSuccess] = useReapitUpdate<undefined, boolean>({
    reapitConnectBrowserSession,
    action: updateActions[UpdateActionNames.updateMember],
    method: 'DELETE',
    uriParams: {
      developerId: member.developerId,
      memberId: member.id,
    },
  })

  const [, memberReinviting, reinviteMember, reinviteMemberSuccess] = useReapitUpdate<InviteMemberModel, boolean>({
    reapitConnectBrowserSession,
    action: updateActions[UpdateActionNames.inviteMember],
    method: 'POST',
    uriParams: {
      developerId: member.developerId,
    },
  })

  useEffect(handleRefreshMembers(refreshMembers, updateMemberSuccess, deleteMemberSuccess, reinviteMemberSuccess), [
    updateMemberSuccess,
    deleteMemberSuccess,
    reinviteMemberSuccess,
  ])

  const isLoading = memberDeleting || memberUpdating || memberReinviting
  const isDisabled = isLoading || member.email === currentMemberEmail

  return (
    <ButtonGroup alignment="center">
      <Button
        intent="primary"
        disabled={isDisabled}
        loading={isLoading}
        onClick={handleUpdateMember(updateMember, {
          ...member,
          role: member.role === 'admin' ? 'user' : 'admin',
        })}
      >
        Set As {member.role === 'admin' ? 'User' : 'Admin'}
      </Button>
      <Button
        intent="primary"
        disabled={isDisabled}
        loading={isLoading}
        onClick={handleReinviteMember(reinviteMember, member, connectSession)}
      >
        Invite Again
      </Button>
      {!member.isMainContact && (
        <Button
          intent="primary"
          disabled={isLoading}
          loading={isLoading}
          onClick={handleUpdateMember(updateMember, {
            ...member,
            isMainContact: true,
          })}
        >
          Set As Main Contact
        </Button>
      )}
      {member.status === 'active' && (
        <Button
          intent="danger"
          disabled={isDisabled}
          loading={isLoading}
          onClick={handleUpdateMember(updateMember, {
            ...member,
            status: 'inactive',
          })}
        >
          Disable
        </Button>
      )}
      <Button intent="danger" disabled={isDisabled} loading={isLoading} onClick={handleDeleteMember(deleteMember)}>
        Delete
      </Button>
    </ButtonGroup>
  )
}
