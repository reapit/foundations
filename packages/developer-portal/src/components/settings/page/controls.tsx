import React, { Dispatch, FC, SetStateAction, useEffect } from 'react'
import {
  Button,
  ButtonGroup,
  elMb7,
  FormLayout,
  Icon,
  InputGroup,
  SmallText,
  Subtitle,
  useModal,
} from '@reapit/elements'
import { useReapitConnect } from '@reapit/connect-session'
import { reapitConnectBrowserSession } from '../../../core/connect-session'
import { useGlobalState } from '../../../core/use-global-state'
import { SendFunction, useReapitUpdate } from '@reapit/use-reapit-data'
import { UpdateActionNames, updateActions } from '@reapit/use-reapit-data'
import { InviteMemberModel } from '@reapit/foundations-ts-definitions'
import { yupResolver } from '@hookform/resolvers/yup'
import { validationSchemaMember } from './validation-schema'
import { useForm } from 'react-hook-form'
import { InputWrapFull } from '@reapit/elements'
import { useLocation } from 'react-router'
import Routes from '../../../constants/routes'

export const handleLogout = (connectLogoutRedirect: () => void) => () => {
  connectLogoutRedirect()
}

export const handleInviteMember =
  (reinviteMember: SendFunction<InviteMemberModel, boolean>, closeModal: () => void) => (member: InviteMemberModel) => {
    reinviteMember(member)
    closeModal()
  }

export const handleInviteMemberSuccess =
  (membersShouldRefresh: Dispatch<SetStateAction<boolean>>, inviteMemberSuccess?: boolean) => () => {
    if (inviteMemberSuccess) {
      membersShouldRefresh(true)
    }
  }

export const Controls: FC = () => {
  const location = useLocation()
  const { connectLogoutRedirect, connectSession } = useReapitConnect(reapitConnectBrowserSession)
  const { globalDataState, globalRefreshState } = useGlobalState()
  const { Modal, openModal, closeModal } = useModal()
  const { currentMember } = globalDataState
  const { members } = globalRefreshState
  const { pathname } = location

  const [, membersShouldRefresh] = members

  const [, , inviteMember, inviteMemberSuccess] = useReapitUpdate<InviteMemberModel, boolean>({
    reapitConnectBrowserSession,
    action: updateActions[UpdateActionNames.inviteMember],
    method: 'POST',
    uriParams: {
      developerId: currentMember?.developerId,
    },
  })

  useEffect(handleInviteMemberSuccess(membersShouldRefresh, inviteMemberSuccess), [inviteMemberSuccess])

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<InviteMemberModel>({
    resolver: yupResolver(validationSchemaMember),
    defaultValues: {
      name: '',
      email: '',
      jobTitle: '',
      message: '',
      sender: connectSession?.loginIdentity.email,
    },
  })

  return (
    <div className={elMb7}>
      <Icon icon="reapitConnectInfographic" iconSize="large" />
      <Subtitle>Settings</Subtitle>
      <SmallText hasGreyText>
        You can manage the information we hold about you and change your password in these pages.
      </SmallText>
      {currentMember?.role === 'admin' && (
        <SmallText hasGreyText>
          Additionally, as an admin, you can manage your developer organisation, members and subscriptions.
        </SmallText>
      )}
      <ButtonGroup>
        {currentMember?.role === 'admin' && pathname === Routes.SETTINGS_MEMBERS && (
          <Button onClick={openModal} intent="primary">
            Invite Developer
          </Button>
        )}
        <Button onClick={handleLogout(connectLogoutRedirect)} intent="primary">
          Logout
        </Button>
      </ButtonGroup>
      <Modal title="Invite Developer">
        <form onSubmit={handleSubmit(handleInviteMember(inviteMember, closeModal))}>
          <FormLayout hasMargin>
            <InputWrapFull>
              <InputGroup
                {...register('name')}
                label="Name"
                placeholder="Your colleague's full name"
                errorMessage={errors?.name?.message}
                icon={errors?.name?.message ? 'asteriskSystem' : null}
                intent="danger"
              />
            </InputWrapFull>
            <InputWrapFull>
              <InputGroup
                {...register('email')}
                type="email"
                label="Email"
                placeholder="An email address we can contact your colleague at"
                errorMessage={errors?.email?.message}
                icon={errors?.email?.message ? 'asteriskSystem' : null}
                intent="danger"
              />
            </InputWrapFull>
            <InputWrapFull>
              <InputGroup
                {...register('jobTitle')}
                label="Job Title"
                placeholder="Your colleague's role at your company'"
                errorMessage={errors?.jobTitle?.message}
                icon={errors?.jobTitle?.message ? 'asteriskSystem' : null}
                intent="danger"
              />
            </InputWrapFull>
            <InputWrapFull>
              <InputGroup
                {...register('message')}
                label="Invite Message"
                placeholder="An optional invite message"
                errorMessage={errors?.message?.message}
                icon={errors?.message?.message ? 'asteriskSystem' : null}
                intent="danger"
              />
            </InputWrapFull>
          </FormLayout>
          <ButtonGroup alignment="right">
            <Button intent="default" onClick={closeModal}>
              Close
            </Button>
            <Button intent="primary" type="submit">
              Send Invite
            </Button>
          </ButtonGroup>
        </form>
      </Modal>
    </div>
  )
}
