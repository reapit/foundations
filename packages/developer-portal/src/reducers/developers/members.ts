import { Action } from '../../types/core'
import { isType } from '../../utils/actions'
import {
  fetchOrganisationMembers,
  fetchOrganisationMembersSuccess,
  fetchOrganisationMembersFailed,
} from '@/actions/developers'
import { PagedResultMemberModel_ } from '@reapit/foundations-ts-definitions'

export interface MembersState {
  loading: boolean
  pagedResult: PagedResultMemberModel_ | null
}

export const defaultState: MembersState = {
  loading: false,
  pagedResult: null,
}

const membersReducer = (state: MembersState = defaultState, action: Action<any>): MembersState => {
  if (isType(action, fetchOrganisationMembers)) {
    return {
      ...state,
      loading: true,
    }
  }
  if (isType(action, fetchOrganisationMembersSuccess)) {
    return {
      ...state,
      pagedResult: action.data,
      loading: false,
    }
  }
  if (isType(action, fetchOrganisationMembersFailed)) {
    return {
      ...state,
      loading: false,
    }
  }

  return state
}

export default membersReducer
