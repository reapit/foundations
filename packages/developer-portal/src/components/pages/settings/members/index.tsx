import React, { FC, useState } from 'react'
import { elMb11, Loader, Pagination, StatusIndicator, Table, Title } from '@reapit/elements'
import { useReapitGet } from '@reapit/utils-react'
import { MemberModelPagedResult } from '@reapit/foundations-ts-definitions'
import { reapitConnectBrowserSession } from '../../../../core/connect-session'
import { GetActionNames, getActions } from '@reapit/utils-common'
import { useReapitConnect } from '@reapit/connect-session'
import { MemberUpdateControls } from './member-update-controls'

export const SettingsMembersPage: FC = () => {
  const { connectSession } = useReapitConnect(reapitConnectBrowserSession)
  const [pageNumber, setPageNumber] = useState<number>(1)
  const developerId = connectSession?.loginIdentity.developerId

  const [members, membersLoading, , refreshMembers] = useReapitGet<MemberModelPagedResult>({
    reapitConnectBrowserSession,
    action: getActions(window.reapit.config.appEnv)[GetActionNames.getDeveloperMembers],
    queryParams: { pageSize: 12, pageNumber },
    uriParams: { developerId },
    fetchWhenTrue: [developerId],
  })

  return (
    <>
      <Title>Members</Title>
      {membersLoading && <Loader />}
      <Table
        className={elMb11}
        rows={members?.data?.map((member) => ({
          cells: [
            {
              label: 'Name',
              value: member.name ?? '',
              icon: 'usernameSystem',
              cellHasDarkText: true,
              narrowTable: {
                showLabel: true,
              },
            },
            {
              label: 'Status',
              value: '',
              children: (
                <>
                  <StatusIndicator
                    intent={
                      member.status === 'active'
                        ? 'success'
                        : member.status === 'rejected'
                        ? 'danger'
                        : member.status === 'pending'
                        ? 'critical'
                        : 'low'
                    }
                  />{' '}
                  {member.status
                    ? `${member.status.charAt(0).toUpperCase()}${member.status.slice(1).toLowerCase()}`
                    : ''}
                </>
              ),
              narrowTable: {
                showLabel: true,
              },
            },
            {
              label: 'Email',
              value: member.email ?? '',
              icon: 'emailSystem',
              cellHasDarkText: true,
              narrowTable: {
                showLabel: true,
              },
            },
            {
              label: 'Role',
              value: member.role ? `${member.role.charAt(0).toUpperCase()}${member.role.slice(1).toLowerCase()}` : '',
              cellHasDarkText: true,
              narrowTable: {
                showLabel: true,
              },
            },
          ],
          expandableContent: {
            content: <MemberUpdateControls member={member} refreshMembers={refreshMembers} />,
          },
        }))}
      />
      <Pagination
        callback={setPageNumber}
        currentPage={pageNumber}
        numberPages={Math.ceil((members?.totalCount ?? 1) / (members?.pageSize ?? 1))}
      />
    </>
  )
}

export default SettingsMembersPage
