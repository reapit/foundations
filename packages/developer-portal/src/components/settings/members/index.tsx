import React, { Dispatch, FC, SetStateAction, useEffect, useState } from 'react'
import {
  Button,
  ButtonGroup,
  elMb11,
  Icon,
  Loader,
  MobileControls,
  Pagination,
  StatusIndicator,
  Table,
  Title,
  useMediaQuery,
  useModal,
} from '@reapit/elements'
import { useReapitGet } from '@reapit/use-reapit-data'
import { Marketplace } from '@reapit/foundations-ts-definitions'
import { reapitConnectBrowserSession } from '../../../core/connect-session'
import { GetActionNames, getActions } from '@reapit/use-reapit-data'
import { useReapitConnect } from '@reapit/connect-session'
import { MemberUpdateControls } from './member-update-controls'
import { useGlobalState } from '../../../core/use-global-state'
import { Controls } from '../page/controls'

export const getIntentFromStatus = (status: string) => {
  return status === 'active'
    ? 'success'
    : status === 'rejected'
      ? 'danger'
      : status === 'pending'
        ? 'pending'
        : 'default'
}

export const handleRefreshMembers =
  (
    membersShouldRefresh: boolean,
    setMembersShouldRefresh: Dispatch<SetStateAction<boolean>>,
    refreshMembers: () => void,
  ) =>
  () => {
    if (membersShouldRefresh) {
      refreshMembers()
      setMembersShouldRefresh(false)
    }
  }

export const SettingsMembersPage: FC = () => {
  const { connectSession } = useReapitConnect(reapitConnectBrowserSession)
  const { globalRefreshState } = useGlobalState()
  const { isMobile } = useMediaQuery()
  const { Modal, openModal, closeModal } = useModal()
  const {
    members: [membersShouldRefresh, setMembersShouldRefresh],
  } = globalRefreshState
  const [pageNumber, setPageNumber] = useState<number>(1)
  const developerId = connectSession?.loginIdentity.developerId

  const [members, membersLoading, , refreshMembers] = useReapitGet<Marketplace.MemberModelPagedResult>({
    reapitConnectBrowserSession,
    action: getActions[GetActionNames.getDeveloperMembers],
    queryParams: { pageSize: 12, pageNumber },
    uriParams: { developerId },
    fetchWhenTrue: [developerId],
  })

  useEffect(handleRefreshMembers(membersShouldRefresh, setMembersShouldRefresh, refreshMembers), [membersShouldRefresh])

  return (
    <>
      <Title>Members</Title>
      {isMobile && (
        <Modal title="Controls">
          <Controls />
          <ButtonGroup alignment="right">
            <Button intent="default" onClick={closeModal}>
              Close
            </Button>
          </ButtonGroup>
        </Modal>
      )}
      {membersLoading && <Loader />}
      <Table
        className={elMb11}
        rows={members?.data?.map((member) => ({
          cells: [
            {
              label: 'Name',
              value: member.name ?? '',
              icon: 'usernameSystem',
              cellHasDarkText: true,
              narrowTable: {
                showLabel: true,
              },
            },
            {
              label: 'Status',
              value: '',
              children: (
                <>
                  <StatusIndicator intent={getIntentFromStatus(member.status ?? '')} />{' '}
                  {`${(member.status ?? '').charAt(0).toUpperCase()}${(member.status ?? '').slice(1).toLowerCase()}`}
                </>
              ),
              narrowTable: {
                showLabel: true,
              },
            },
            {
              label: 'Email',
              value: member.email ?? '',
              icon: 'emailSystem',
              cellHasDarkText: true,
              narrowTable: {
                showLabel: true,
              },
            },
            {
              label: 'Role',
              value: member.role ? `${member.role.charAt(0).toUpperCase()}${member.role.slice(1).toLowerCase()}` : '',
              cellHasDarkText: true,
              narrowTable: {
                showLabel: true,
              },
            },
            {
              label: 'Main Contact',
              value: member.isMainContact ? <Icon icon="check" intent="success" /> : '-',
              narrowTable: {
                showLabel: true,
              },
            },
          ],
          expandableContent: {
            content: <MemberUpdateControls member={member} refreshMembers={refreshMembers} />,
          },
        }))}
      />
      <Pagination
        callback={setPageNumber}
        currentPage={pageNumber}
        numberPages={Math.ceil((members?.totalCount ?? 1) / (members?.pageSize ?? 1))}
      />
      <MobileControls onClick={openModal} />
    </>
  )
}

export default SettingsMembersPage
