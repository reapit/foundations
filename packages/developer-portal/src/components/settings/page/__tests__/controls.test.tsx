import React from 'react'
import { render } from '../../../../tests/react-testing'
import { mockMemberModel } from '../../../../tests/__stubs__/members'
import { Controls, handleInviteMember, handleInviteMemberSuccess, handleLogout } from '../controls'

jest.mock('../../../../core/use-global-state')

describe('Controls', () => {
  it('should match a snapshot', () => {
    expect(render(<Controls />)).toMatchSnapshot()
  })
})

describe('handleLogout', () => {
  it('should handle member update', () => {
    const connectLogoutRedirect = jest.fn()

    const curried = handleLogout(connectLogoutRedirect)

    curried()

    expect(connectLogoutRedirect).toHaveBeenCalledTimes(1)
  })
})

describe('handleInviteMember', () => {
  it('should handle member update', () => {
    const inviteMember = jest.fn()
    const closeModal = jest.fn()
    const member = mockMemberModel

    const curried = handleInviteMember(inviteMember, closeModal)

    curried(member)

    expect(inviteMember).toHaveBeenCalledWith(member)
  })
})

describe('handleInviteMemberSuccess', () => {
  it('should handle member update', () => {
    const membersShouldRefresh = jest.fn()
    const inviteMemberSuccess = true
    const curried = handleInviteMemberSuccess(membersShouldRefresh, inviteMemberSuccess)

    curried()

    expect(membersShouldRefresh).toHaveBeenCalledWith(true)
  })
})
