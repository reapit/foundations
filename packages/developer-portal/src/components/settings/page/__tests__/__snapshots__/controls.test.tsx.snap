// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls should match a snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <mock-styled.div />
      <div
        class="mock-css.
  margin-bottom: 1.25rem;
"
      >
        <mock-styled.span
          classname="mock-css.
  font-size: 5rem;
"
        >
          <svg />
        </mock-styled.span>
        <mock-styled.h2
          classname=""
        >
          Settings
        </mock-styled.h2>
        <mock-styled.p
          classname="mock-css."
        >
          You can manage the information we hold about you and change your password in these pages.
        </mock-styled.p>
        <mock-styled.p
          classname="mock-css."
        >
          Additionally, as an admin, you can manage your developer organisation, members and subscriptions.
        </mock-styled.p>
        <mock-styled.div>
          <mock-styled.div
            classname=""
          >
            <mock-styled.button
              classname="mock-css."
            >
              <mock-styled.div />
              Logout
            </mock-styled.button>
          </mock-styled.div>
        </mock-styled.div>
      </div>
    </div>
  </body>,
  "container": <div>
    <mock-styled.div />
    <div
      class="mock-css.
  margin-bottom: 1.25rem;
"
    >
      <mock-styled.span
        classname="mock-css.
  font-size: 5rem;
"
      >
        <svg />
      </mock-styled.span>
      <mock-styled.h2
        classname=""
      >
        Settings
      </mock-styled.h2>
      <mock-styled.p
        classname="mock-css."
      >
        You can manage the information we hold about you and change your password in these pages.
      </mock-styled.p>
      <mock-styled.p
        classname="mock-css."
      >
        Additionally, as an admin, you can manage your developer organisation, members and subscriptions.
      </mock-styled.p>
      <mock-styled.div>
        <mock-styled.div
          classname=""
        >
          <mock-styled.button
            classname="mock-css."
          >
            <mock-styled.div />
            Logout
          </mock-styled.button>
        </mock-styled.div>
      </mock-styled.div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Controls should match a snapshot for the members page 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <mock-styled.div />
      <div
        class="mock-css.
  margin-bottom: 1.25rem;
"
      >
        <mock-styled.span
          classname="mock-css.
  font-size: 5rem;
"
        >
          <svg />
        </mock-styled.span>
        <mock-styled.h2
          classname=""
        >
          Settings
        </mock-styled.h2>
        <mock-styled.p
          classname="mock-css."
        >
          You can manage the information we hold about you and change your password in these pages.
        </mock-styled.p>
        <mock-styled.p
          classname="mock-css."
        >
          Additionally, as an admin, you can manage your developer organisation, members and subscriptions.
        </mock-styled.p>
        <mock-styled.div>
          <mock-styled.div
            classname=""
          >
            <mock-styled.button
              classname="mock-css."
            >
              <mock-styled.div />
              Logout
            </mock-styled.button>
          </mock-styled.div>
        </mock-styled.div>
      </div>
    </div>
  </body>,
  "container": <div>
    <mock-styled.div />
    <div
      class="mock-css.
  margin-bottom: 1.25rem;
"
    >
      <mock-styled.span
        classname="mock-css.
  font-size: 5rem;
"
      >
        <svg />
      </mock-styled.span>
      <mock-styled.h2
        classname=""
      >
        Settings
      </mock-styled.h2>
      <mock-styled.p
        classname="mock-css."
      >
        You can manage the information we hold about you and change your password in these pages.
      </mock-styled.p>
      <mock-styled.p
        classname="mock-css."
      >
        Additionally, as an admin, you can manage your developer organisation, members and subscriptions.
      </mock-styled.p>
      <mock-styled.div>
        <mock-styled.div
          classname=""
        >
          <mock-styled.button
            classname="mock-css."
          >
            <mock-styled.div />
            Logout
          </mock-styled.button>
        </mock-styled.div>
      </mock-styled.div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
